### PR TITLE
Fix stripping backslash from new line characters in JS files (#200)

### PR DESCRIPTION
**Pull request by jakubjo**
This is the PR for previously created issue #200.

I'm not quite sure if its enough.
Everything went fine in my tests with 756 strings accross 89 files.

### CODE REVIEW

File: src/Utils/JsFunctionsScanner.php [hunk 1]
```diff
@@ -37,11 +37,18 @@ public function getFunctions(array $constants = [])
 
             switch ($char) {
                 case '\\':
-                    $prev = $char;
-                    $char = $next;
                     $pos++;
+
+                    if ($next !== 'n') {
```
Comment (oscarotero) [inline src/Utils/JsFunctionsScanner.php]: The escaped character cannot be only `\n` but any character (like, for example, `\t`).
I think we should accept any character inside double or single quotes.

Comment (jakubjo) [reply src/Utils/JsFunctionsScanner.php]: I'll look into it. So essentially we could just remove the whole `\\` case from the switch statement?

Comment (oscarotero) [reply src/Utils/JsFunctionsScanner.php]: I guess just adding something like this:

```php
switch ($this->status()) {
    case 'simple-quote':
    case 'double-quote':
        break;
    default:
        //everything else
}
```

**Comment by jakubjo** [comment]
Fixed it -- tests are running fine.

Screenshot 2019-03-25 00 29 53

**Comment by jakubjo** [comment]
Unfortunately we've encountered additional issues where js files are not parser properly.
I'll investigate further and come back with more informations later.

**Comment by jakubjo** [comment]
I've migrated the JS parser to use @mck89's peast parser.

The parser had issues with various, presumably legacy styled, files. Some strings would not be extracted. 
I've added a failing test first, so you're able to verify the bug by checking out commit bc85e89.

## Alle tests are now green. 
Additionally I've tested it on 285 different JS files ranging from legacy ugly pre ES2015 styled files to newest babel based styled files.

**Comment by oscarotero** [comment]
Sorry but I wouldn't like to add more dependencies to this package, specially if it's something required only by one extractor (and most users wont use).

If you want to use `peast` (and looks a great project), I recomend to create a different package for this extractor and use Gettext as dependency.

**Comment by jakubjo** [comment]
The elegance of your package lies in the all-in-one solution.

Why care about an additional dependency? Is there any significant issue with an additional dependency? That's the soul of open source: Don't reinvent the wheel. 

> and most users wont use

But essentially you don't know. The extractor is part of your package and has a bug in the current implementation. 
Which would be fixed with this PR.

**So final words were spoken?**

**Comment by oscarotero** [comment]
> The elegance of your package lies in the all-in-one solution.

To me, the elegance of this package is that it can be extended by the community with other features. For example, there are twig translators, symfony bundlers, laravel and cake integrations, wordpress, cli tools, etc. I love that people want to create and maintain these packages, but I wouldn't like to have it integrated in this package. The more dependencies and features, the more work and time is needed to maintain everything up to date.

The extractors provided do not require dependencies, because many people install this package only for the po->mo conversion, or any other specific feature, so I do not want to force them to install a package that they do not need, or even worse, may cause conflicts with their dependencies.

I made a commit to solve this issue. Let me know if it works fine now.

**Comment by jakubjo** [comment]
Thank you very much for your efforts. Its perfect! 👍

**Comment by oscarotero** [comment]
Great! I close this.

**Pull request closed without merging**